scalableminds/webknossos
# Better diffs

Adapt our diffing algorithm (or better the compactUpdateActions function) to detect tree merges and splits to:

  - avoid  extremely large JSON requests to the server (that fail if too big)
  - avoid broken tracings due to those big requests
  - increase overall tree merge/split performance, which takes a couple of minutes currently for a 30000 node tree

As I've written in the issue, merging a 30000 node tree is now down from >5m to <5s.

I'm not entirely happy with this currently as I think the code is kind of complicated and may be error prone. I've added comments where appropriate, but the code is not really polished. If you have improvement suggestions, please tell me!

Also this is very important code and should be thoroughly tested, otherwise we'll have to deal with broken tracings!

### Steps to test:
- Run the tests
- Open a tracing (try with a big nml if you have one), split and merge trees, reload the page -> everything should still be there and work
- Time to split/merge big trees (>10000 nodes) should not be bigger than a couple seconds

### Issues:
- fixes #1857 and the broken tracings issue that comes up from time to time

------

[4500nodes.nml.zip](https://github.com/scalableminds/webknossos/files/1052466/4500nodes.nml.zip)
[30000_nodes.nml.zip](https://github.com/scalableminds/webknossos/files/1052467/30000_nodes.nml.zip)
